### ISSUE TITLE
Question: is this equivalent?

### ISSUE THREAD

**Issue by enumag** [amphp/amp#289]
I'm still relatively new to amp and async PHP so I'm trying to get a better understanding of how promises work. I just noticed that you can resolve a `Deferred` instance with a promise though I'm not entirely sure how that works. I have a hypotheses though. Are these two snippets of code equivalent?

```php
// asyncOperation() returns Promise

$deferred = new Deferred();
$deferred->resolve(asyncOperation());
```

```php
$deferred = new Deferred();
asyncOperation()->onResolve(
	function (?Throwable $exception, $value) use ($deferred): void {
		if ($exception !== null) {
			$deferred->fail($exception);
		} else {
			$deferred->resolve($value);
		}
	}
);
```

**Comment by kelunik** [comment]
They're nearly equivalent, but not 100%. You can find the actual resolution code [here](https://github.com/amphp/amp/blob/dd4d6be6778e8968c2c0a47465ad2d156201cbfc/lib/Internal/Placeholder.php#L83-L148).

The difference is that the first example immediately marks the deferred as resolved, so future calls to `$deferred->resolve()` or `$deferred->fail()` will immediately throw an exception, while the second example resolves the deferred only after the async operation finished, so anything calling `$deferred->resolve()` or `$deferred->fail()` in between will win and the call inside `asyncOperation()->onResolve()` will throw an exception instead.

**Comment by enumag** [comment]
I see. Thanks for the detailed explanation!
